### ISSUE TITLE
Flatten type class structures.

### ISSUE THREAD

**Issue by sellout** [FormationAI/dhall-bhat#8]
Right now we have structures that look like
```dhall
{ functor : { map : ... }
, applicative : { ap : ..., pure : ... }
, monad : { bind : ... }
}
```
but these should be flattened to
```dhall
{ map : ...
, ap : ...
, pure : ...
, bind : ...
}
```
with functions that project out super-instances from sub-instances. With new features like record type combining (`//\\`) and multi-field projection, this will be easier.

As is commonly done in Scala, we should be able to define
```dhall
{ ...
, bind = ...
, traverse = ...
} : Monad ./Foo //\\ Traversable ./Foo
```
in one fell swoop.

**Comment by FintanH** [comment]
I can't find where we can do `//\\`?

I do see a way of doing this though. If we have the structure:
```
-- Functor/map
λ(f : Type → Type)
→ { ∀(a : Type) → ∀(b : Type) → ∀(g : a → b) → ∀(fa : f a) → f b }
```
```
-- Functor/Type
λ(f : Type → Type) → { map : ./map f }
```
```
-- Applicative/pure
λ(f : Type → Type) → ∀(a : Type) → a → f a
```
```
-- Applicative/ap
λ(f : Type → Type) → ∀(a : Type) → ∀(b : Type) → ∀(g : f (a → b)) → ∀(fa : f a) → f b
```
```
-- Applicative/Type
λ(f : Type -> Type) ->
            { ap   : ./ap f
            , pure : ./pure f
            , map : ./Functor/map f
            }
```

**Comment by sellout** [comment]
> With new features like record type combining (`//\\`) and multi-field projection, this will be easier.

FYI, these features were _just_ released in Dhall 1.13 (https://twitter.com/dhall_lang/status/990252437528178691), so we’re not able to take advantage of them until that version is more widely available and we can officially upgrade Platform to it. I would say, once it’s in a nixpkgs that we’ve pinned to?

**Comment by coreykanon** [comment]
On hold - waiting for Dhall 1.14 release - needs to be Platform-wide update to newer version.
Newer version supports cleaner refactoring - but not a strict requirement.
Nice to Have.
